### ISSUE TITLE
Multi-class advice support

### ISSUE THREAD

**Issue by orange-buffalo** [raphw/byte-buddy#145]
Hello,

This is an enhancement request for `Advice` API. It would be very useful if `Advice` transformation could look for `@OnMethodEnter` and `@OnMethodExit` not only in the class specified for `.to(..)` method, but also in the superclasses of this class. This would help better code organization where multiple advices share the same enter/exit functionality. 

For instance, consider code like this
```
class BaseAdvice {
    @Advice.OnMethodEnter
    public static void before() {
        // common functionality to execute before method body
    }
}

class FirstAdvice extends BaseAdvice {
    @Advice.OnMethodExit
    public static void after() {
        // specific code to run after method body execution
    }
}

class SecondAdvice extends BaseAdvice {
    @Advice.OnMethodExit
    public static void after() {
        // specific code to run after method body execution
    }
}
```

And then calls to 
```
Advice.to(FirstAdvice.class).on(...)
...
Advice.to(SecondAdvice.class).on(...)
```
would enable both `@OnMethodEnter` and `@OnMethodExit`.

Alternative solution would be to allow fine tuning of Advice, for example
```
Advice.to(FirstAdvice.class).onMethodExit().to(BaseAdvice.class).onMethodEnter().on(...)
...
Advice.to(SecondAdvice.class).onMethodExit().to(BaseAdvice.class).onMethodEnter().on(...)
```

**Comment by CodingFabian** [comment]
I like the idea. I however see practical problems when adviced methods call super(). That basically would mean that advice code would trigger twice for an invocation which conceptually should trigger only once.

Your proposal would make sure that with crazy subclass combinations you are unlikely to miss an invocation, but on the other hand are likely to double count.
The double counting is not avoidable by instrumentation, but would be needed to be checked by advice logic. (aka: if called by method which has been adviced by this advice, do not execute advice code again)

We have such recursion control in our use case already, and it would be nice to have native byte buddy support for it.
However, I am a bit sceptical if BB could offer a usable API which fits all use cases. I feel that it is still better as user to control this.

**Comment by raphw** [comment]
Generally, I understand why this is desirable. One practical problem is however the performance implication. Currently, the `Advice` class shares one `ClassReader` for parsing advice byte code from the same class. If the advice would be distributed among several classes, this optimization would no longer be possible.

Out of curiosity; what stops you from registering two advice adapters?

```java
.visit(Advice.to(BaseAdvice.class).on(...))
.visit(Advice.to(FirstAdvice.class).on(...))
```

This should have the same effect.

**Comment by orange-buffalo** [comment]
As far as I understand, if I have the same matchers in `on(..)` for two advice adapters, the matchers' code will be executed twice for every transformed class. This influences performance when matchers are more or less complicated.
Another problem is that with advanced Advice configuration (like custom bindings) there will be code duplication and/or code readability will be reduced.

**Comment by raphw** [comment]
You are right about the matchers, one way to prevent this would be the use of `ElementMatchers::cached`. As for the custom bindings, you can however share the configuration. All `Advice` instances are immutable such that you can simply continue the advice creation after setting it up once.

I will look into this and see if there is a way of doing this without hurting performance.

**Comment by raphw** [comment]
I added an alternative entry point to the advice component which allows you to specify two different classes for an entry and an exit advice: `Advice.to(EntryAdvice.class, ExitAdvice.class)`

I prefer this over parsing the class hierarchy as the latter could yield ambiguities and I assume that a user typically can provide the information of where to find an advice class.

Furthermore, I refactored the `AsmVisitorWrapper.ForDeclaredMethods` to allow reusing a matcher:
```java
new AsmVisitorWrapper.ForDeclaredMethods()
  .method(matcher, Advice.to(Foo.class), Advice.to(Bar.class))
```

As a matter of fact, `Advice.to(Foo.class).on(matcher)` was never anything more than a short-cut for the above construct anyways.

**Comment by orange-buffalo** [comment]
Super, thank you very much!
